themes/default/content/blog/kubernetes-v4-shared-resource-management-with-server-side-apply/index.md

```diff
@@ -0,0 +1,219 @@
+
+Today, we are excited to release the latest major version of our Kubernetes provider - Pulumi Kubernetes 4.0.
+
+<!--more-->
+
+Pulumi Kubernetes 4.0 uses Kubernetes Server-Side Apply by default, supports upserting on all resources, and enables new Patch resources for every resource type in the provider.  It also brings improved diffs, removes dependence on kubectl annotations, and provides simpler access to outputs in the Python, Go and Java SDKs, and with any Kubernetes cluster version >= 1.13 and in any self-hosted or managed Kubernetes cluster.
+
+Together, these additions make Pulumi’s support for Kuberntes even more Kubernetes-native, allowing the best of both worlds between Kubernetes’ flexible resource model and Pulumi’s robust desired state infrastructure as code model, all while using your favorite languages (TypeScript, Python, Go, .NET and Java) to define and manage your Kubernetes Infrastructure.  Pulumi programs that need to manage Kubernetes resources that are also modified by other systems (external controllers, managed Kubernetes service control planes, kubectl or other IaC solutions) can faithfully represent the modifications they want to make to existing resources within the Pulumi resource lifecycle.  And Pulumi’s diffs will represent the accurate differences based on the current state of resources in the cluster.
+
+![Upsert](https://www.pulumi.com/blog/kubernetes-server-side-apply/upsert.gif)
+
+## Server-Side Apply, Upserts and Patch Resources
+
+Kubernetes [Server-side Apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) (SSA) mode is now default in the v4 provider. SSA mode enables several useful new features: Kubernetes resources can be safely managed by more than one controller. Any conflicting fields will produce a detailed error message which avoids accidentally overriding changes made by other controllers. Resources can be “upserted”; in other words, create the resource if it does not exist, or apply the specified configuration to an existing resource. Patch resources add a new way to manage shared Kubernetes resources. Patch resources can be used to make targeted changes, and do not require the entire resource to be specified. This is a convenient way to update shared resources; a common example is resources that are provisioned by the cloud provider rather than as part of the Pulumi stack. Deleting a Patch resource undoes the patched change, but does not remove the shared Kubernetes resource. See the [SSA how-to guide](https://www.pulumi.com/registry/packages/kubernetes/how-to-guides/managing-resources-with-server-side-apply/) for more details and usage examples for SSA mode.
+
+> Note: Users may opt into the previous default Client-side Apply (CSA) mode by setting the `enableServerSideApply` provider configuration to “false”.
+
+{{< chooser language "typescript,python,go,csharp,java,yaml" >}}
```

Comment:
This example is for upserting a resource. We should add a heading. It might also be worth adding the Patch example from https://www.pulumi.com/registry/packages/kubernetes/how-to-guides/managing-resources-with-server-side-apply/#patch-a-resource

Reply:
Added a note - but still could potentially be nice to include a second code example.  Though in context, I think it is okay to have only one inlined here - and to leave the rest for the migration guide.